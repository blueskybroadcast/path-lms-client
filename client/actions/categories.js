import normalize from 'jsonapi-normalizer';

import {
  FETCH_CATEGORIES_SUCCESS, DELETE_CATEGORY_SUCCESS, FETCH_CATEGORIES_REQUEST,
  ADD_CATEGORY_REQUEST, ADD_CATEGORY_SUCCESS, ADD_CATEGORY_FAILURE
} from './types';

export const fetchCategories = () => async (dispatch, getState, api) => {
  dispatch({ type: FETCH_CATEGORIES_REQUEST });
  const res = await api.get('/categories');

  dispatch({
    type: FETCH_CATEGORIES_SUCCESS,
    payload: normalize(res.data)
  });
};

export const addCategory = category => async (dispatch, getState, api) => {
  dispatch({ type: ADD_CATEGORY_REQUEST });

  const { name, parentId, hidden } = category;

  await api.post('/categories', {
    name,
    parent_id: parentId,
    hidden
  }).then(() => {
    dispatch({ type: ADD_CATEGORY_SUCCESS });
    dispatch(fetchCategories());
  }).catch(() => {
    dispatch({ type: ADD_CATEGORY_FAILURE });
  });
};

export const deleteCategory = id => async (dispatch, getState, api) => {
  await api.delete(`/categories/${id}`).then(() => {
    dispatch({
      type: DELETE_CATEGORY_SUCCESS,
      payload: id
    });
  });
};

export const editCategory = (id, category) => async (dispatch, getState, api) => {
  const { name, parentId, hidden } = category;

  await api.put(`/categories/${id}`, {
    name,
    parent_id: parentId,
    hidden
  }).then(() => {
    dispatch(fetchCategories());
  });
};
