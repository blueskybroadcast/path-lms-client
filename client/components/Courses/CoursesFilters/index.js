import LibraryFilters from './container';

export default LibraryFilters;
