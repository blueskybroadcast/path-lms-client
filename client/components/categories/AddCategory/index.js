import AddCategory from './container';

export default AddCategory;
