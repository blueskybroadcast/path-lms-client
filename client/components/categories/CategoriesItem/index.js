import CategoriesItem from './container';

export default CategoriesItem;
