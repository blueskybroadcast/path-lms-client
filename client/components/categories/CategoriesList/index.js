import CategoriesList from './container';

export default CategoriesList;
