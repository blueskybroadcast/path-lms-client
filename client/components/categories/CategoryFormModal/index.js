import CategoryFormModal from './container';

export default CategoryFormModal;
