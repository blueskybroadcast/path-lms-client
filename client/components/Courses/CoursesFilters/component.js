import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Link } from 'react-router-dom';
import queryString from 'query-string';
import { isEmpty } from 'lodash';

import CoursesFiltersCategories from './components/CoursesFiltersCategories';
import { getCategoriesForSelect } from '../../../helpers/categories';

const initialState = {
  categories: [],
  filter: undefined,
  search: ''
};

class CoursesFilters extends React.Component {
  state = initialState

  componentDidMount() {
    const { location } = this.props;
    const options = queryString.parse(location.search, { arrayFormat: 'bracket' });
    const opts = {
      categories: options.category_ids || initialState.categories,
      filter: options.filter,
      search: options.search || initialState.search
    };
    this.setState(opts);
  }

  accountHasFeature = (feature) => {
    const { currentAccountFeatures } = this.props;
    return currentAccountFeatures.indexOf(feature) !== '-1';
  }

  fetchCourses = () => {
    const { slug, history, fetchCourses } = this.props;
    const { categories, filter, search } = this.state;

    const options = {
      category_ids: categories,
      search: search || undefined,
      filter
    };

    history.push({
      pathname: `/${slug}/courses`,
      search: queryString.stringify(options, { arrayFormat: 'bracket' })
    });

    fetchCourses(options);
  }

  handleSearchChange = ({ target: { value } }) => {
    this.setState({ search: value });

    clearTimeout(this.delayTimer);
    this.delayTimer = setTimeout(() => {
      this.fetchCourses();
    }, 1000);
  }

  handleCategorySelect = (id) => {
    const { categories } = this.state;
    if (categories.indexOf(id) === -1) {
      this.setState(
        () => ({ categories: [...categories, id] }),
        () => this.fetchCourses()
      );
    }
  }

  handleCategoryRemove = (id) => {
    const { categories } = this.state;
    categories.splice(categories.indexOf(id), 1);
    this.setState(
      () => ({ categories }),
      () => this.fetchCourses()
    );
  }

  handleCategoriesReset = () => {
    this.setState({ categories: [] });
  }

  handleFilterChange = (filter) => {
    this.setState(
      () => ({ filter }),
      () => this.fetchCourses()
    );
  }

  renderFilter = (name, id) => (
    <li>
      <a
        className={classNames('filter cursor-pointer', this.state.filter === id && 'active')}
        onClick={() => this.handleFilterChange(id)}
      >
        {name}
      </a>
    </li>
  )

  delayTimer

  render() {
    const {
      isAdmin, slug, categoriesIds, categoriesData
    } = this.props;
    const { search, categories } = this.state;

    const categoriesForSelect = getCategoriesForSelect(categoriesIds, categoriesData);
    const filtersAreAvailable = this.accountHasFeature('ecommerce')
      && this.accountHasFeature('sellable_filters');
    const accountHasFeatureFreeCle = this.accountHasFeature('free_cle');

    return (
      <div className="filtering-panel">
        <div className="filtering-controls">
          <div
            className={classNames('categories-list', isAdmin && 'with-manage-button')}
          >
            <CoursesFiltersCategories
              categories={categoriesForSelect}
              handleCategorySelect={this.handleCategorySelect}
              handleCategoriesReset={this.handleCategoriesReset}
            />

            { isAdmin && (
              <Link
                className="manage-categories tooltiped"
                to={`/${slug}/categories`}
              >
                <i className="icon icon-cog" />
              </Link>
            )}
          </div>
          <form className="search-courses-form" data-role="search-form">
            <input
              value={search}
              onChange={this.handleSearchChange}
              placeholder="Search in programs..."
              type="text"
            />
          </form>
          { filtersAreAvailable && (
            <ul className="filters-list">
              { this.renderFilter('All') }
              { !accountHasFeatureFreeCle && this.renderFilter('Free', 'free') }
              { accountHasFeatureFreeCle && this.renderFilter('Free CLE', 'free_cle') }
              { !accountHasFeatureFreeCle && (
                <>
                  { this.renderFilter('Purchased', 'purchased') }
                  { this.renderFilter('Not Purchased', 'not_purchased') }
                </>
              )}
            </ul>
          )}
        </div>
        { categories.length && !isEmpty(categoriesData) ? (
          <ul className="tags-list" data-role="chosen-categories">
            { categories.map(id => (
              <li className="tag-wrapper" key={id}>
                <a
                  onClick={() => this.handleCategoryRemove(id)}
                  className="tag-item cursor-pointer"
                >
                  {categoriesData[id].name}
                  <i className="icon icon-times" />
                </a>
              </li>
            ))}
          </ul>
        ) : null}
      </div>
    );
  }
}

export default CoursesFilters;
