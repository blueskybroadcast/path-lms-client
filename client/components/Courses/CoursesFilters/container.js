import { connect } from 'react-redux';

import { currentAccountSlugSelector, currentUserIsAdmin } from '../../../selectors/authSelectors';

import CoursesFilters from './component';

const mapStateToProps = state => ({
  isAdmin: currentUserIsAdmin(state),
  slug: currentAccountSlugSelector(state)
});

export default connect(mapStateToProps)(CoursesFilters);
