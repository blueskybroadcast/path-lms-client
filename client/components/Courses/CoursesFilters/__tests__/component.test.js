import React from 'react';
import { shallow } from 'enzyme';
import CoursesFilters from '../component';

import { categoriesIds, categoriesData } from '../../../../fixtures/categories';

describe('Courses filters', () => {
  test('should render CoursesFilters with data and admin rights correctly', () => {
    const search = '';
    const wrapper = shallow(
      <CoursesFilters
        location={{ search }}
        categoriesIds={categoriesIds}
        categoriesData={categoriesData}
        currentAccountFeatures={[]}
        slug="testing"
        history={{ push: jest.fn() }}
        fetchCourses={jest.fn()}
        isAdmin
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should render CoursesFilters without data and admin rights, with search term correctly', () => {
    const search = '?search=asdasd';
    const wrapper = shallow(
      <CoursesFilters
        location={{ search }}
        currentAccountFeatures={[]}
        slug="testing"
        history={{ push: jest.fn() }}
        fetchCourses={jest.fn()}
        isAdmin={false}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
