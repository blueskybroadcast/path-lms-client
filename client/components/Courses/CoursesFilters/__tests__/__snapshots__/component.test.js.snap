// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Courses filters should render CoursesFilters with data and admin rights correctly 1`] = `
<div
  className="filtering-panel"
>
  <div
    className="filtering-controls"
  >
    <div
      className="categories-list with-manage-button"
    >
      <CoursesFiltersCategories
        categories={
          Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "hidden": false,
                  "id": "30",
                  "name": "Another Level 0 child",
                },
              ],
              "hidden": false,
              "id": "28",
              "name": "Another Level 0",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "hidden": false,
                      "id": "27",
                      "name": "Level 0 child child",
                    },
                  ],
                  "hidden": false,
                  "id": "26",
                  "name": "Level 0 child",
                },
              ],
              "hidden": false,
              "id": "25",
              "name": "Level 0",
            },
            Object {
              "children": Array [],
              "hidden": false,
              "id": "29",
              "name": "Super Level 0",
            },
          ]
        }
        handleCategoriesReset={[Function]}
        handleCategorySelect={[Function]}
      />
      <Link
        className="manage-categories tooltiped"
        replace={false}
        to="/testing/categories"
      >
        <i
          className="icon icon-cog"
        />
      </Link>
    </div>
    <form
      className="search-courses-form"
      data-role="search-form"
    >
      <input
        onChange={[Function]}
        placeholder="Search in programs..."
        type="text"
        value=""
      />
    </form>
    <ul
      className="filters-list"
    >
      <li>
        <a
          className="filter cursor-pointer active"
          onClick={[Function]}
        >
          All
        </a>
      </li>
      <li>
        <a
          className="filter cursor-pointer"
          onClick={[Function]}
        >
          Free CLE
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Courses filters should render CoursesFilters without data and admin rights, with search term correctly 1`] = `
<div
  className="filtering-panel"
>
  <div
    className="filtering-controls"
  >
    <div
      className="categories-list"
    >
      <CoursesFiltersCategories
        categories={Array []}
        handleCategoriesReset={[Function]}
        handleCategorySelect={[Function]}
      />
    </div>
    <form
      className="search-courses-form"
      data-role="search-form"
    >
      <input
        onChange={[Function]}
        placeholder="Search in programs..."
        type="text"
        value="asdasd"
      />
    </form>
    <ul
      className="filters-list"
    >
      <li>
        <a
          className="filter cursor-pointer active"
          onClick={[Function]}
        >
          All
        </a>
      </li>
      <li>
        <a
          className="filter cursor-pointer"
          onClick={[Function]}
        >
          Free CLE
        </a>
      </li>
    </ul>
  </div>
</div>
`;
